File: backend/app/api/v1/endpoints/groups.py
import uuid

from fastapi import APIRouter, HTTPException, status

from app.api.deps import AuthDep, DbDep
from app.schemas.groups import GroupCreate, GroupDetailRead, GroupRead, GroupUpdate
from app.services.groups import GroupConflictError, GroupNotFoundError, GroupService

router = APIRouter(prefix="/groups", tags=["groups"])


def _require_admin(current_user: AuthDep) -> None:
    if not current_user.has_role("platform_admin", "org_admin"):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Insufficient permissions")


@router.get("", response_model=list[GroupRead], summary="List all groups")
async def list_groups(current_user: AuthDep, db: DbDep) -> list[GroupRead]:
    if not current_user.has_role("platform_admin", "org_admin", "auditor"):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Insufficient permissions")
    return await GroupService(db).list_groups()


@router.post("", response_model=GroupDetailRead, status_code=status.HTTP_201_CREATED, summary="Create group")
async def create_group(body: GroupCreate, current_user: AuthDep, db: DbDep) -> GroupDetailRead:
    _require_admin(current_user)
    try:
        return await GroupService(db).create_group(body)
    except GroupConflictError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc)) from exc


@router.get("/{group_id}", response_model=GroupDetailRead, summary="Get group details")
async def get_group(group_id: uuid.UUID, current_user: AuthDep, db: DbDep) -> GroupDetailRead:
    if not current_user.has_role("platform_admin", "org_admin", "auditor"):
        raise HTTPException(status_code=status.HTTP_403_FORBIDDEN, detail="Insufficient permissions")
    try:
        return await GroupService(db).get_group(group_id)
    except GroupNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Group not found") from exc


@router.patch("/{group_id}", response_model=GroupDetailRead, summary="Update group")
async def update_group(group_id: uuid.UUID, body: GroupUpdate, current_user: AuthDep, db: DbDep) -> GroupDetailRead:
    _require_admin(current_user)
    try:
        return await GroupService(db).update_group(group_id, body)
    except GroupNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Group not found") from exc
    except GroupConflictError as exc:
        raise HTTPException(status_code=status.HTTP_409_CONFLICT, detail=str(exc)) from exc


@router.delete("/{group_id}", status_code=status.HTTP_204_NO_CONTENT, summary="Delete group")
async def delete_group(group_id: uuid.UUID, current_user: AuthDep, db: DbDep) -> None:
    _require_admin(current_user)
    try:
        await GroupService(db).delete_group(group_id)
    except GroupNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Group not found") from exc


@router.post("/{group_id}/members/{user_id}", status_code=status.HTTP_204_NO_CONTENT, summary="Add member to group")
async def add_member(group_id: uuid.UUID, user_id: uuid.UUID, current_user: AuthDep, db: DbDep) -> None:
    _require_admin(current_user)
    try:
        await GroupService(db).add_member(group_id, user_id)
    except GroupNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Group not found") from exc


@router.delete("/{group_id}/members/{user_id}", status_code=status.HTTP_204_NO_CONTENT, summary="Remove member from group")
async def remove_member(group_id: uuid.UUID, user_id: uuid.UUID, current_user: AuthDep, db: DbDep) -> None:
    _require_admin(current_user)
    try:
        await GroupService(db).remove_member(group_id, user_id)
    except GroupNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Group not found") from exc


@router.post(
    "/{group_id}/apply-roles",
    summary="Grant group roles to all current members",
    response_model=dict,
)
async def apply_roles(group_id: uuid.UUID, current_user: AuthDep, db: DbDep) -> dict:
    _require_admin(current_user)
    try:
        count = await GroupService(db).apply_roles_to_members(group_id, granted_by=current_user.user_id)
        return {"grants_added": count}
    except GroupNotFoundError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail="Group not found") from exc

File: frontend/src/shared/api/groups.ts
import { apiClient } from "./client";
import type { UserRole } from "./users";

export interface GroupMember {
  id: string;
  email: string;
  display_name: string;
  is_active: boolean;
}

export interface Group {
  id: string;
  name: string;
  description: string | null;
  created_at: string;
  member_count: number;
  roles: UserRole[];
}

export interface GroupDetail extends Group {
  members: GroupMember[];
}

export interface GroupCreate {
  name: string;
  description?: string;
  role_ids: string[];
}

export interface GroupUpdate {
  name?: string;
  description?: string;
  role_ids?: string[];
}

export async function listGroups(): Promise<Group[]> {
  const { data } = await apiClient.get<Group[]>("/groups");
  return data;
}

export async function getGroup(groupId: string): Promise<GroupDetail> {
  const { data } = await apiClient.get<GroupDetail>(`/groups/${groupId}`);
  return data;
}

export async function createGroup(body: GroupCreate): Promise<GroupDetail> {
  const { data } = await apiClient.post<GroupDetail>("/groups", body);
  return data;
}

export async function updateGroup(groupId: string, body: GroupUpdate): Promise<GroupDetail> {
  const { data } = await apiClient.patch<GroupDetail>(`/groups/${groupId}`, body);
  return data;
}

export async function deleteGroup(groupId: string): Promise<void> {
  await apiClient.delete(`/groups/${groupId}`);
}

export async function addGroupMember(groupId: string, userId: string): Promise<void> {
  await apiClient.post(`/groups/${groupId}/members/${userId}`);
}

export async function removeGroupMember(groupId: string, userId: string): Promise<void> {
  await apiClient.delete(`/groups/${groupId}/members/${userId}`);
}

export async function applyGroupRoles(groupId: string): Promise<{ grants_added: number }> {
  const { data } = await apiClient.post<{ grants_added: number }>(`/groups/${groupId}/apply-roles`);
  return data;
}
